### DISCUSSION TITLE
Possibility of output object with named values?

### DISCUSSION THREAD

**Discussion by xsvato01** [nextflow-io/nextflow#5499] [Q&A] [answered]
Hi, I was wondering if it is possible in Nextflow to ouput named object from processes. For example:

```
process Foo {
	output:
	tuple path("file.txt"), ['word': env(SAY), 'something_else': val(8)]

	script:
	"""
        touch file.txt
	SAY=\$(echo whatever)
 	"""
}
```
This will not work, but something similar can be achieved with
```
process Foo {
	output:
	tuple path("file.txt"), env(SAY),  val(8)

	script:
	"""
        touch file.txt
	SAY=\$(echo whatever)
 	"""
}

workflow {
	Foo()
	.map{ it -> [ it[0], [word: it[1], something_else: it[2] ] }
	.view()
}
```

Perhaps there is any more elegant way without needing to map over the individual outputs...?

**Comment by bentsherman** [comment, marked as answer]
For now your example is the best you can do. Named outputs (record types) are on the roadmap, but there are a few moving pieces involved so I'm not yet sure when it will be ready

**Reply by xsvato01** [reply]
Lovely, thanks for the heads-up!

**Reply by tobsecret** [reply]
This is very exciting! Will this also come with support for named arguments?

Have been looking over a bunch of `nf-core` pipelines recently.
I cannot shake the feeling that a lack of support for named arguments and named outputs for processes is what leads to patterns that I find hard to read.

```
some_earlier_process_out_ch.map{
it -> [it[0], it[2], it[3]] 
}.set{
some_ch
}
SOME_PROCESS(some_ch)
```

I now need to go back to that earlier process and read what its outputs were and in which order they occur.
Maybe I'm also not making enough use of DSL2 functionality and this is already sth that can be done? Would love to hear advice!

**Reply by bentsherman** [reply]
You can already do a sort of tuple de-structuring with closures:
```groovy
some_earlier_process_out_ch
  .map { a, b, c, d -> [a, c, d] }
  .set { some_ch }
SOME_PROCESS(some_ch)
```

**Reply by tobsecret** [reply]
Ah nice, that's helpful! Thank you so much for the quick followup!
I still think we would like being able to pass values between processes entirely by name but this is def helpful in the interims - so much more readable.

**Reply by tobsecret** [reply]
So the way I am thinking about it, it would look sth like:
```
SOME_PROCESS(a = some_earlier_process_out_ch.a, c = some_earlier_process_out_ch.c, d = some_earlier_process_out_ch.d)
```
Or even simpler, if the accepted arg names are in the hashmap that is passed:
```
SOME_PROCESS(some_earlier_process_out_ch)
```
Similar to python's `some_func(**kwargs)` pattern.
